rpc/server.go

```diff
@@ -17,54 +20,81 @@ var log = logging.Logger("rpc")
 
 // Server represents the JSON-RPC server
 type Server struct {
-	s    *rpc.Server
-	port uint16
+	ctx        context.Context
+	listener   net.Listener
+	httpServer *http.Server
 }
 
-// Config ...
+// Config is the RPC server configuration passed to NewServer
 type Config struct {
-	Port    uint16
+	Ctx     context.Context
+	Address string // "IP:port" to listen on
 	Relayer *relayer.Relayer
 }
 
-// NewServer ...
+// NewServer creates but does not start the relayer RPC service. Pre-start configuration
+// that can fail, including reserving the server port, is handled here.
 func NewServer(cfg *Config) (*Server, error) {
-	s := rpc.NewServer()
-	s.RegisterCodec(NewCodec(), "application/json")
+	rpcServer := rpc.NewServer()
+	rpcServer.RegisterCodec(NewCodec(), "application/json")
 
-	rs, err := NewRelayerService(cfg.Relayer)
+	relayService, err := NewRelayerService(cfg.Relayer)
 	if err != nil {
 		return nil, err
 	}
 
-	if err := s.RegisterService(rs, "relayer"); err != nil {
+	if err := rpcServer.RegisterService(relayService, "relayer"); err != nil {
+		return nil, err
+	}
+
+	addr := cfg.Address
+	if addr == "" {
+		// default to restricted localhost access with a randomized port
+		addr = "127.0.0.1:0"
+	}
+
+	router := mux.NewRouter()
+	router.Handle("/", rpcServer)
+
+	headersOk := handlers.AllowedHeaders([]string{"content-type", "username", "password"})
+	methodsOk := handlers.AllowedMethods([]string{"GET", "HEAD", "POST", "PUT", "OPTIONS"})
+	originsOk := handlers.AllowedOrigins([]string{"*"})
+
+	lc := net.ListenConfig{}
+	ln, err := lc.Listen(cfg.Ctx, "tcp", addr)
+	if err != nil {
 		return nil, err
 	}
 
 	return &Server{
-		s:    s,
-		port: cfg.Port,
+		ctx:      cfg.Ctx,
+		listener: ln,
+		httpServer: &http.Server{
+			Addr:              ln.Addr().String(),
+			ReadHeaderTimeout: time.Second,
+			Handler:           handlers.CORS(headersOk, methodsOk, originsOk)(router),
+			BaseContext: func(listener net.Listener) context.Context {
+				return cfg.Ctx
+			},
+		},
 	}, nil
 }
 
+// HttpURL returns the URL used for HTTP requests
+func (s *Server) HttpURL() string { //nolint:revive
+	return fmt.Sprintf("http://%s", s.httpServer.Addr)
+}
+
 // Start starts the JSON-RPC server.
 func (s *Server) Start() <-chan error {
-	errCh := make(chan error)
+	errCh := make(chan error, 1)
```

Comment:
I thought it was slightly better to let the go routine that generates the error complete instead of blocking.